CONTRIBUTING.md

```diff
@@ -1,76 +1,401 @@
+and copy the current work tree until it is a specified size.
+
+Test Your Changes on Linux
+--------------------------
+
+It can be important to work directly on the [core Git codebase](https://github.com/git/git),
+such as a recent commit into the `master` or `next` branch that has not been incorporated
+into Git for Windows. Also, it can help to run functional and performance tests on your
+code in Linux before submitting patches to the Linux-focused mailing list.
+
+### Using the Windows Subsystem for Linux
+
+The [Windows Subsystem for Linux (WSL)](https://docs.microsoft.com/en-us/windows/wsl/install-win10)
+allows you to [install Ubuntu Linux as an app](https://www.microsoft.com/en-us/store/p/ubuntu/9nblggh4msv6)
+that can run Linux executables on top of the Windows kernel. Internally,
+Linux syscalls are interpreted by the WSL, everything else is plain Ubuntu.
+
+First, open WSL (either type "Bash" in Cortana, or execute "bash.exe" in a CMD window).
+Then install the prerequisites, and `git` for the initial clone:
+
+```
+sudo apt-get update
+sudo apt-get install git gcc make libssl-dev libcurl4-openssl-dev libexpat-dev tcl tk gettext git-email
+```
+
+Then, clone and build:
+
+```
+git clone https://github.com/git-for-windows/git
+cd git
+git remote add -f upstream https://github.com/git/git
+make
+```
+
+Be sure to clone into `/home/[user]/` and not into any folder under `/mnt/?/` or your build
+will fail due to colons in file names.
+
+### Using a Linux Virtual Machine with Hyper-V
+
+If you prefer, you can use a virtual machine (VM) to run Linux and test your changes in the
+full environment. I like this setup, particularly because the test suite runs a lot faster
+on Linux than on Windows or with the WSL. I connect to the VM using an SSH terminal like
+[PuTTY](http://www.putty.org/). I've tried to make the following steps be as clear as
+possible, without missing any steps.
+
+* [Download an Ubuntu Server ISO](https://www.ubuntu.com/download/server).
+* Open [Hyper-V Manager](https://docs.microsoft.com/en-us/virtualization/hyper-v-on-windows/quick-start/enable-hyper-v).
+* [Set up a virtual switch](https://docs.microsoft.com/en-us/virtualization/hyper-v-on-windows/quick-start/connect-to-network)
+  so your VM can reach the network.
+* Select "Quick Create", name your machine, select the ISO as installation source, and un-check
+  "This virtual machine will run Windows."
+* Go through the Ubuntu install process, being sure to select to install OpenSSH Server.
+* When install is complete, log in and check the SSH server status with `sudo service ssh status`.
+    * If the service is not found, install with `sudo apt-get install openssh-server`.
+    * If the service is not running, then use `sudo service ssh start`.
+* Use `shutdown -h now` to shutdown the VM, go to the Hyper-V settings for the VM, expand Network Adapter
+  to select "Advanced Features", and set the MAC address to be static (this can save your VM from losing

```

Comment:
>set the MAC address to be static

Surely, you mean "IP address?" or does Hyper-V defaults to change MAC addresses?

Reply:
I do mean MAC address. This issue hit me and I had to mess with some crazy stuff to get SSH working again.

Reply:
ahm. ok. That's... special.

Reply:
Please also notice that recommending the Client version of Ubuntu will make those `sshd` steps fail. Might be better to keep suggesting the Server version.

Reply:
`sudo apt-get install openssh-server` works fine on ubuntu-client versions.

Reply:
That's an extra step. The idea, I thought, was to keep the instructions as simple as possible? Windows users may not even *want* to be educated in things Ubuntu, but merely get their patches tested on Linux, too.

Reply:
My initial comment was that it's easier to install ubuntu-client than ubuntu-server, especially for non-linux-savvy users. Also, `install openssh-server` shows up in line 251.